[Pytorch][Ondevice quantization] Add device side API to convert model

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84131
* __->__ #83807
* #83742
* #83571
* #83570
* #83569
* #83568

Summary:
This diff adds device side API which will convert the model to its
quantized equivalent. THe input model must have been prepared AOT for
quantization.

API is implemented by:
- Running reset obervers
- Running observe method
- Running quantize method
- And replacing method, e.g. forward, with its quantized equivalent.

Test Plan:
test/quantization/jit/test_ondevice_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D38889818](https://our.internmc.facebook.com/intern/diff/D38889818)

kimishpatel committed:
[Pytorch][Ondevice quantization] Add device side API to convert model

Summary:
This diff adds device side API which will convert the model to its
quantized equivalent. THe input model must have been prepared AOT for
quantization.

API is implemented by:
- Running reset obervers
- Running observe method
- Running quantize method
- And replacing method, e.g. forward, with its quantized equivalent.

Test Plan:
test/quantization/jit/test_ondevice_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38889818).

kimishpatel committed:
Update on "[Pytorch][Ondevice quantization] Add device side API to convert model"


Summary:
This diff adds device side API which will convert the model to its
quantized equivalent. THe input model must have been prepared AOT for
quantization.

API is implemented by:
- Running reset obervers
- Running observe method
- Running quantize method
- And replacing method, e.g. forward, with its quantized equivalent.

Test Plan:
test/quantization/jit/test_ondevice_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D38889818](https://our.internmc.facebook.com/intern/diff/D38889818)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38889818).

iseeyuan commented on test/quantization/jit/test_ondevice_quantization.py:
```diff
@@ -467,6 +468,66 @@ def _check_serialization_deserialization(self, model):
+        ref_output = ref_m(*inputs)
+
+        # check for lite interpreter
+        m = OnDevicePTQUtils.ptq_dynamic_quantize(model, qconfig_dict)
```
Dumb question: What's the difference between `m` and `ref_m`? Any reason we need two interfaces? Is it because that there are additional functions (`quantized_forward`, `quantize_forward`, `observe_forward`, `reset_observes_forward`)?

iseeyuan commented on test/quantization/jit/test_ondevice_quantization.py:
```diff
@@ -467,6 +468,66 @@ def _check_serialization_deserialization(self, model):
+        output = m(*inputs)
+        self.assertTrue(torch.allclose(ref_output, output))
+
+        model.eval()
+        inputs = model.get_example_inputs()
+        ref_m = torch.jit.script(model)
+        torch._C._jit_pass_inline(ref_m.graph)
+        qconfig_dict = {"" : per_channel_dynamic_qconfig}
+        ref_m = prepare_dynamic_jit(ref_m, qconfig_dict)
+        ref_m = convert_dynamic_jit(ref_m)
+        buffer = io.BytesIO()
+        torch.jit.save(ref_m, buffer)
+        buffer.seek(0)
+        ref_m = torch.jit.load(buffer)
+        ref_output = ref_m(*inputs)
+
+        # check for lite interpreter
+        m = OnDevicePTQUtils.ptq_dynamic_quantize(model, qconfig_dict)
+        first_input, = inputs
+        rand_input = bundled_inputs.bundle_randn(first_input.size(), dtype=first_input.dtype)
+        m = bundled_inputs.bundle_inputs(m, inputs=[(rand_input, )])
+        buffer = io.BytesIO(m._save_to_buffer_for_lite_interpreter())
+        buffer.seek(0)
+        m = _load_for_lite_interpreter(buffer)  # Error here
+        torch._C._quantize_ondevice_ptq_dynamic(m._c, "forward")
+        self.assertFalse(m.find_method("quantized_forward"))
+        self.assertFalse(m.find_method("quantize_forward"))
+        self.assertFalse(m.find_method("observe_forward"))
+        self.assertFalse(m.find_method("reset_observers_forward"))
+        output = m(*inputs)
+        self.assertTrue(torch.allclose(ref_output, output))
```
Any reason we run the similar codes twice? Could we put it into a function?

kimishpatel replied:
Sure, I can combine.

iseeyuan commented on torch/csrc/jit/mobile/quantization.h:
```diff
@@ -0,0 +1,38 @@
+ * Device side PTQ API.
+ * Once the model has been prepared for quantization on server side, such model
+ * is sent to device. On device side the model is further trained. At the end of
+ * the training, before the model is readied for inference, we need to quantize
```
```suggestion
 * the training, before the model is ready for inference, we need to quantize
```

iseeyuan commented on torch/csrc/jit/mobile/quantization.cpp:
```diff
@@ -0,0 +1,64 @@
+      m.find_method("get_all_bundled_inputs"),
+      "PTQ ready module must have get_all_bundled_inputs method.");
+
+  auto inputs = m.run_method("get_all_bundled_inputs")
```
Another dumb question: if the inputs are bundled aot, any reason we cannot replace the quantized method (i.e., call `PTQQuanizationHelper::quantize_dynamic`) aot? Which parts can only be observed at runtime?

kimishpatel replied:
No we cannot do aot because mode is not yet trained. Once the model is trained then only it is ready to be quanitzed. The training happens on the device and model does not leave the device.

iseeyuan replied:
I see. I think I missed this context. Thanks!

iseeyuan commented on build_variables.bzl:
```diff
@@ -562,6 +562,7 @@ torch_mobile_core = [
     "torch/csrc/jit/mobile/observer.cpp",
     "torch/csrc/jit/mobile/parse_bytecode.cpp",
     "torch/csrc/jit/mobile/parse_operators.cpp",
+    "torch/csrc/jit/mobile/quantization.cpp",
```
What's the size effect of this file?

kimishpatel replied:
It is quite small actually and also note that only here it is in mobile_core. In buckbuild.bzl i put it under training dep (no equivalent of that in build_variables.bzl), so in reality it should not have any impact on binary size.

kimishpatel committed:
Update on "[Pytorch][Ondevice quantization] Add device side API to convert model"


Summary:
This diff adds device side API which will convert the model to its
quantized equivalent. THe input model must have been prepared AOT for
quantization.

API is implemented by:
- Running reset obervers
- Running observe method
- Running quantize method
- And replacing method, e.g. forward, with its quantized equivalent.

Test Plan:
test/quantization/jit/test_ondevice_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D38889818](https://our.internmc.facebook.com/intern/diff/D38889818)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38889818).

kimishpatel committed:
Update on "[Pytorch][Ondevice quantization] Add device side API to convert model"


Summary:
This diff adds device side API which will convert the model to its
quantized equivalent. THe input model must have been prepared AOT for
quantization.

API is implemented by:
- Running reset obervers
- Running observe method
- Running quantize method
- And replacing method, e.g. forward, with its quantized equivalent.

Test Plan:
test/quantization/jit/test_ondevice_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D38889818](https://our.internmc.facebook.com/intern/diff/D38889818)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38889818).

iseeyuan reviewed (approved):
Synced offline. Looks like is one-time training + quantization that is required from some solid use case. Since it does not introduce much overhead of MobileModule, it looks good to me.

kimishpatel committed:
Update on "[Pytorch][Ondevice quantization] Add device side API to convert model"


Summary:
This diff adds device side API which will convert the model to its
quantized equivalent. THe input model must have been prepared AOT for
quantization.

API is implemented by:
- Running reset obervers
- Running observe method
- Running quantize method
- And replacing method, e.g. forward, with its quantized equivalent.

Test Plan:
test/quantization/jit/test_ondevice_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D38889818](https://our.internmc.facebook.com/intern/diff/D38889818)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38889818).

kimishpatel committed:
Update on "[Pytorch][Ondevice quantization] Add device side API to convert model"


Summary:
This diff adds device side API which will convert the model to its
quantized equivalent. THe input model must have been prepared AOT for
quantization.

API is implemented by:
- Running reset obervers
- Running observe method
- Running quantize method
- And replacing method, e.g. forward, with its quantized equivalent.

Test Plan:
test/quantization/jit/test_ondevice_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D38889818](https://our.internmc.facebook.com/intern/diff/D38889818)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38889818).

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38889818).

kimishpatel committed:
Update on "[Pytorch][Ondevice quantization] Add device side API to convert model"


Summary:
This diff adds device side API which will convert the model to its
quantized equivalent. THe input model must have been prepared AOT for
quantization.

API is implemented by:
- Running reset obervers
- Running observe method
- Running quantize method
- And replacing method, e.g. forward, with its quantized equivalent.

Test Plan:
test/quantization/jit/test_ondevice_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D38889818](https://our.internmc.facebook.com/intern/diff/D38889818)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38889818).

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38889818).
